kart/cli.py

```diff
@@ -318,6 +322,63 @@ def lfs(ctx, args):
     execvp("git", [*params, *args])
 
 
+@cli.command(
+    name="ext-run", context_settings=dict(ignore_unknown_options=True), hidden=True
+)
+@click.pass_context
+@click.argument("script", type=click.Path(exists=True, dir_okay=False), required=True)
+@click.argument("args", nargs=-1, type=click.UNPROCESSED)
+def ext_run(ctx, script, args):
+    """
+    Invoke the main() method from an arbitrary Python script in the Kart
+    environment. There is no public API for Kart.
+    """
+    script = pathlib.Path(script)
+    # strip all the suffixes
+    module_name = "kart.ext_run." + script.name.split(".", 1)[0]
+
+    def _print_ext_tb():
+        # print a traceback from the ext module down
+        extracts = traceback.extract_tb(sys.exc_info()[2])
+        count = len(extracts)
+        # find the first occurrence of the module file name
+        for i, extract in enumerate(extracts):
+            if extract[0] == script.name:
+                break
+            count -= 1
+        # keep only the count of last lines
+        traceback.print_exc(limit=-count)
+
+    spec = importlib.util.spec_from_file_location(module_name, script)
+    module = importlib.util.module_from_spec(spec)
+    try:
+        spec.loader.exec_module(module)
+    except Exception:
+        click.secho(f"Error loading {script}:", fg="red", file=sys.stderr)
+        _print_ext_tb()
+        raise click.Abort()
+
+    if not hasattr(module, "main") or not callable(module.main):
+        click.secho(
+            f"{script} does not have a main() function", fg="red", file=sys.stderr
+        )
+        raise click.Abort()
+
+    try:
+        f_sig = inspect.signature(module.main)
+        f_sig.bind(ctx=ctx, args=args)
+    except TypeError:
+        click.secho(
+            f"Invalid function definition for {script}:main()",
+            fg="red",
+            file=sys.stderr,
+        )
+        _print_ext_tb()
+        raise click.Abort()
+
+    return module.main(ctx=ctx, args=args)
```

Comment:
I think we should enable `main` to be a `click.command`; atm this doesn't work:

```python
import click


@click.command()
@click.pass_context
def main(ctx, args):
    print(args)
```


results in:

```python
$ kart ext-run x.py
Traceback (most recent call last):
...
  File "/Users/cdestigter/checkout/kart/kart/cli.py", line 379, in ext_run
    return module.main(ctx=ctx, args=args)
  File "/Users/cdestigter/checkout/kart/venv/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/cdestigter/checkout/kart/venv/lib/python3.7/site-packages/click/core.py", line 781, in main
    with self.make_context(prog_name, args, **extra) as ctx:
  File "/Users/cdestigter/checkout/kart/venv/lib/python3.7/site-packages/click/core.py", line 698, in make_context
    ctx = Context(self, info_name=info_name, parent=parent, **extra)
TypeError: __init__() got an unexpected keyword argument 'ctx'
```

Reply:
This would require inspecting to see if `main` was decorated as a click command in the `ext-run` code as extensions could add `@click.option` and `@click.argument` and so get named or positional parameters to the main function, correct?

It would be nice to be able to do that and potentially have a group of commands but I think that would require 'registering' the extension somehow so it is imported and becomes part of the click hierarchy - something like `kart --ext x.py mycommand --frob==wongle`, where `--ext x.py` just imports that file and doesn't run anything directly itself. A natural path from that is to allow multiple `--ext` options and actually fits better with the forking model for speed ups - it would come for free.

Reply:
I think there's a difference between a plugin (`kart myplugin`) which needs registering into click in advance for help/arg handling; and an extension, which just invokes a script. This is the latter, the former is something we might want in future.

If the extension wants to setup click/argparse/whatever, then good on it.

Reply:
Yep. The distinction is tricky there as either way the code has to fit a particular model - to support click all code must use it etc. and likewise for no click support it must do it itself, plug-ins give more scope to make opinionated choices on capability.

Reply:
I passed the click context purely because some stuff that might have already been parsed globally could be useful: verbosity, `-c git.config.var=x`, whatever. Wasn't really for _actual_ use 😄

Reply:
yep, nbd, I guess all that's needed is a tiny wrapper like:

```python

def main(*, ctx, **kw):
    ctx.invoke(actualmain, **kw)
```